flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java

```diff
@@ -1916,6 +1918,46 @@ else if (value instanceof Row) {
 		}
 	}
 
+	public static <X> TypeInformation<X> getForCollection(Iterable<X> value) {
+		return new TypeExtractor().privateGetForIterable(value);
+	}
+
+	public static <X> TypeInformation<X> getForCollection(X[] value) {
+		return new TypeExtractor().privateGetForIterable(Arrays.asList(value));
+	}
+
+	@SuppressWarnings({ "unchecked", "rawtypes" })
+	private <X> TypeInformation<X> privateGetForIterable(Iterable<X> value) {
+		checkNotNull(value);
+
+		Iterator<X> it = value.iterator();
+		X v = checkNotNull(it.next());
+
+		if (v instanceof Row) {
```

Comment:
So this only works if there is one `Row` for which every field is not null? Why not instead search the type of every column separately, and then create the `RowTypeInfo` manually? Then it would only fail if one field is null for every `Row`.